boot/boot_test.go

```diff
@@ -902,3 +902,457 @@ func (s *bootenv20Suite) TestMarkBootSuccessful20BaseUpdate(c *C) {
+	c.Assert(err, IsNil)
+	r := s.bootloader.SetRunKernelImageEnabledKernel(kernel)
+	defer r()
+
+	// make a new kernel
+	kernel2, err := snap.ParsePlaceInfoFromSnapFileName("pc-kernel_2.snap")
+	c.Assert(err, IsNil)
+
+	// default state
+	s.bootloader.BootVars["kernel_status"] = boot.DefaultStatus
+
+	// get the boot kernel participant from our new kernel snap
+	bootKern := boot.Participant(kernel2, snap.TypeKernel, coreDev)
+	// make sure it's not a trivial boot participant
+	c.Assert(bootKern.IsTrivial(), Equals, false)
+
+	// we will panic during SetBootVars
+	restoreBootloaderPanic := s.bootloader.SetRunKernelImagePanic("SetBootVars")
+
+	// attempt set next, expecting a panic in SetBootVars
+	c.Check(
+		func() { bootKern.SetNextBoot() },
+		PanicMatches,
+		"mocked reboot panic in SetBootVars",
+	)
+
+	// don't panic anymore
+	restoreBootloaderPanic()
+
+	// do the bootloader kernel failover logic handling
+	nextBootingKernel, err := runBootloaderLogic(c, s.bootloader)
+	c.Assert(err, IsNil)
+
+	// the next kernel we boot from should be kernel, because we didn't change
+	// the kernel_status
+	c.Assert(nextBootingKernel, Equals, kernel)
+
+	// we should now at least have kernel_status in default state because
+	// SetBootVars completed successfully
+	vars, err := s.bootloader.GetBootVars("kernel_status")
+	c.Assert(err, IsNil)
+	c.Assert(vars, DeepEquals, map[string]string{"kernel_status": boot.DefaultStatus})
+
+	// the modeenv now has this kernel listed however
+	m2, err := boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	c.Assert(m2.CurrentKernels, DeepEquals, []string{"pc-kernel_1.snap", "pc-kernel_2.snap"})
+
+	// now, after a reboot running MarkBootSuccessful should work
+	err = boot.MarkBootSuccessful(coreDev)
+	c.Assert(err, IsNil)
+
+	// after running now, we should still have Kernel() == pc-kernel_1.snap
+	// because nothing was committed
+	afterKernel, err := s.bootloader.Kernel()
+	c.Assert(err, IsNil)
+	c.Assert(afterKernel, DeepEquals, kernel)
+
+	// kernel_status is still default as well
+	vars, err = s.bootloader.GetBootVars("kernel_status")
+	c.Assert(err, IsNil)
+	c.Assert(vars, DeepEquals, map[string]string{"kernel_status": boot.DefaultStatus})
+
+	// we should also still have TryKernel() == pc-kernel_2.snap, as well as
+	// trying == true, because the symlink is still there, the snapstate mgr
+	// is responsible for cleaning it up when it notices the boot with the try
+	// kernel failed - that is done a level above this test
+	afterTryKernel, err := s.bootloader.TryKernel()
+	c.Assert(err, IsNil)
+	c.Assert(afterTryKernel, Equals, kernel2)
+}
+
+// runBootloaderLogic implements the logic from the gadget snap bootloader,
+// namely that we transition kernel_status "try" -> "trying" and "trying" -> ""
+// and use try-kernel.efi when kernel_status is "try" and kernel.efi in all
+// other situations
+func runBootloaderLogic(c *C, ebl bootloader.ExtractedRunKernelImageBootloader) (snap.PlaceInfo, error) {
```

Comment:
we have similar code in other places, we should see later how/if to share via boottest

Reply:
as I hinted this can be taken care of in a follow up

Reply:
Yeah, checking if we could somehow run the real failover logic would be amazing